packages/github/src/github.ts

```diff
@@ -34,125 +25,39 @@ export class GitHub extends Octokit {
-    const defaults: GraphQLRequestParameters = {}
-    defaults.baseUrl = this.getGraphQLBaseUrl()
-    const token = args[0]
-    const options = args[1]
-
-    // Authorization
-    const auth = this.getAuthString(token, options)
-    if (auth) {
-      defaults.headers = {
-        authorization: auth
+  constructor(token: string, opts?: Omit<OctokitOptions, 'auth'>)
+  constructor(opts: OctokitOptions)
+  constructor(token: string | OctokitOptions, opts?: OctokitOptions) {
+    super(Utils.getOctokitOptions(Utils.disambiguate(token, opts)))
+    this.graphql = Utils.getGraphQL(Utils.disambiguate(token, opts))
+  }
+
+  // Base class assumes the octokit options are the first arg, that is not the case for us, so we need to reimplement
+  static defaults<T extends Constructor<any>>(
+    this: T,
+    defaults: OctokitOptions
+  ): T {
+    const GitHubWithDefaults = class extends this {
+      constructor(...args: any[]) {
+        const token = args[0]
+        const opts = args[1]
+        const options = Utils.getOctokitOptions(Utils.disambiguate(token, opts))
+        super(Object.assign({}, defaults, options))
       }
     }
-
-    // Proxy
-    const agent = GitHub.getProxyAgent(defaults.baseUrl, options)
-    if (agent) {
-      defaults.request = {agent}
-    }
-
-    return graphql.defaults(defaults)
-  }
-
-  private static getAuthString(
-    token: string,
-    options: Octokit.Options
-  ): string | undefined {
-    // Validate args
-    if (!token && !options.auth) {
-      throw new Error('Parameter token or opts.auth is required')
-    } else if (token && options.auth) {
-      throw new Error(
-        'Parameters token and opts.auth may not both be specified'
-      )
-    }
-
-    return typeof options.auth === 'string' ? options.auth : `token ${token}`
-  }
-
-  private static getProxyAgent(
-    destinationUrl: string,
-    options: Octokit.Options
-  ): http.Agent | undefined {
-    if (!options.request?.agent) {
-      if (httpClient.getProxyUrl(destinationUrl)) {
-        const hc = new httpClient.HttpClient()
-        return hc.getAgent(destinationUrl)
-      }
-    }
-
-    return undefined
+    return GitHubWithDefaults
   }
+}
 
-  private static getApiBaseUrl(): string {
-    return process.env['GITHUB_API_URL'] || 'https://api.github.com'
-  }
-
-  private static getGraphQLBaseUrl(): string {
-    let url =
-      process.env['GITHUB_GRAPHQL_URL'] || 'https://api.github.com/graphql'
-
-    // Shouldn't be a trailing slash, but remove if so
-    if (url.endsWith('/')) {
-      url = url.substr(0, url.length - 1)
-    }
-
-    // Remove trailing "/graphql"
-    if (url.toUpperCase().endsWith('/GRAPHQL')) {
-      url = url.substr(0, url.length - '/graphql'.length)
-    }
-    return url
+const baseUrl = Utils.getApiBaseUrl()
+const defaults = {
+  baseUrl,
+  request: {
+    agent: Utils.getProxyAgent(baseUrl)
   }
 }
+
+export const GitHub = GitHubCore.plugin(
```

Comment:
We can eventually export an Enterprise version using the enterprise plugins instead of `restEndpointMethods` as well to support GHES

export const GitHubEnterprise = GitHubCore.plugin(paginateRest, require("@octokit/plugin-enterprise-server")).defaults(defaults)